[SandboxVectorizer] Use sbvec-passes flag to create a pipeline of Region passes after BottomUpVec.

slackito opened the pull request:
The main change is that the main SandboxVectorizer pass no longer has a pipeline of function passes. Now it consists of a BottomUpVec pass and a pipeline of region passes after it.

BottomUpVec now takes a RegionPassManager as an argument to the constructor. This argument is currently stored and not used, but the idea is that after BottomUpVec vectorizes a region of a function, we'll run successive optimization passes on that region.

I've also moved creation of the pass pipeline out of the `run` method to the SandboxVectorizer constructor.

slackito committed:
[SandboxVectorizer] Use sbvec-passes flag to create a pipeline of Region passes after BottomUpVec.

The main change is that the main SandboxVectorizer pass no longer has a
pipeline of function passes. Now it consists of a BottomUpVec pass and
a pipeline of region passes after it.

BottomUpVec now takes a RegionPassManager as an argument to the
constructor. This argument is currently stored and not used, but the
idea is that after BottomUpVec vectorizes a region of a function, we'll
run successive optimization passes on that region.

I've also moved creation of the pass pipeline out of the `run` method to
the SandboxVectorizer constructor.

slackito commented:
I'm not sure this is the way it should work but if not at least I hope to get some clarity out of the review :)

slackito committed:
clang-format

vporpo commented on llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.h:
```diff
@@ -17,10 +21,22 @@ class TargetTransformInfo;
+  sandboxir::PassRegistry PR;
+
+  // The main vectorizer pass.
+  std::unique_ptr<sandboxir::BottomUpVec> BottomUpVecPass;
```
Why is this a member variable? It's created and used only within `runOnFunction()` at the moment, so making it a class member implies that it used across multiple member functions of `SandboxVectorizerPass`. Are you planning to move it to a separate class that `SandboxVectorizerPass` inherits from?

aeubanks replied:
if we move away from `cl::opt("sbvec-passes")` and to proper pass parameters, we would want the LLVM pass to hold the region pass manager in the class, rather than keep it local to `runImpl`, because we'd specify the region pass pipeline via the constructor

but with `cl::opt` I don't think it really matters

aeubanks replied:
actually, I thought that `BottomUpVecPass` would own the region pass manager so it could invoke it when it creates a region

slackito replied:
> making it a class member implies that it used across multiple member functions of SandboxVectorizerPass

Not the only reason possible to make something a member. Making it local to `runImpl` means constructing the `BottomUpVec` pass every time, which seems like a waste to me.

Also, conceptually I see the `SandboxVectorizer` pass as a wrapper around `BottomUpVec` that handles the creation of sandbox IR. I don't see anything wrong with the wrapper owning the thing it's wrapping.

vporpo replied:
> Not the only reason possible to make something a member. Making it local to runImpl means constructing the BottomUpVec pass every time, which seems like a waste to me.

Ah yes, I forgot that we were building the pass pipeline over and over again.

aeubanks replied:
> actually, I thought that BottomUpVecPass would own the region pass manager so it could invoke it when it creates a region

I missed `BottomUpVec` also having a reference to the region pass manager

vporpo commented on llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.h:
```diff
@@ -17,10 +21,22 @@ class TargetTransformInfo;
+
+  // The PM containing the pipeline of region passes. It's owned by the pass
+  // registry.
+  sandboxir::RegionPassManager *RPM;
```
Same

vporpo commented on llvm/lib/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.cpp:
```diff
@@ -30,6 +31,57 @@ cl::opt<std::string> UserDefinedPassPipeline(
+  registerAllRegionPasses(PR);
+
+  // Create a pipeline to be run on each Region created by BottomUpVec.
+  if (UserDefinedPassPipeline == DefaultPipelineMagicStr) {
```
Same for this stuff. Not sure why they should be in the pass constructor.

slackito replied:
Why recreate the pipeline for each function we ever run the vectorizer on, if it's defined by a command-line flag and won't change during a compiler invocation?

vporpo replied:
Yeah totally forgot that we were rebuilding it every time.

aeubanks commented on llvm/lib/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.cpp:
```diff
@@ -56,31 +108,13 @@ bool SandboxVectorizerPass::runImpl(Function &LLVMF) {
-    PM = &PR.parseAndCreatePassPipeline(UserDefinedPassPipeline);
-  }
-
   if (PrintPassPipeline) {
```
oh, is this the only reason `SandboxVectorizerPass` has a reference to the region pass manager? I think ideally only the `BoUpVectorizer` has a reference to it, and we can call `BottomUpVecPass->printPipeline()`, which forwards to `RPM->printPipeline()`

slackito replied:
Yeah, I'm fine with `BottomUpVec` owning the region pass manager instead. I considered it and took the lazy way out because it made it easier to print the pipeline here :)

I'll change it.

slackito replied:
Done. I'm not sure I'm completely happy with the pattern where the RegionPassManager object is actually owned by the PassRegistry, but I don't think it will cause any problems either and if we go with it we can always change it afterwards.

aeubanks replied:
yeah, all the non-owning pointers make me sus

aeubanks reviewed (commented):
seems fine conceptually

slackito committed:
Moved the region pass pipeline to BottomUpVec.

Everything related to the pass pipeline is now in BottomUpVec.cpp,
including pipeline creation and printing and the flags to control it.

I have also changed the `BottomUpVecPass` in `SandboxVectorizer` from
a `unique_ptr` to a plain old member, as I don't think we need that
level of indirection for anything.

aeubanks commented on llvm/lib/Transforms/Vectorize/SandboxVectorizer/Passes/BottomUpVec.cpp:
```diff
@@ -10,10 +10,73 @@
+/// A magic string for the default pass pipeline.
+const char *DefaultPipelineMagicStr = "*";
+
+cl::opt<std::string> UserDefinedPassPipeline(
```
this can be static/anonymous namespace

slackito committed:
Make BottomUpVec own its RegionPassManager.

Also made some flag-related globals static.

slackito commented:
> yeah, all the non-owning pointers make me sus

Got rid of the non-owning pointer. Now the region pass manager is a simple field of `BottomUpVec`.

vporpo commented on llvm/lib/Transforms/Vectorize/SandboxVectorizer/Passes/BottomUpVec.cpp:
```diff
@@ -10,10 +10,69 @@
+
+/// Adds to `RPM` a sequence of passes described by `Pipeline` from the `PR`
+/// pass registry.
+static void parseAndCreatePassPipeline(RegionPassManager &RPM, PassRegistry &PR,
```
Is this function going to be used for non RegionPasses too? If not, should it be a member function of RPM?

slackito replied:
It could be a member of RPM. The moral equivalent of this function for `FunctionPassManager` is currently a member function of `PassRegistry`:
```
class PassRegistry {
  ...
  FunctionPassManager &parseAndCreatePassPipeline(StringRef Pipeline);
};
```

If we change `PassRegistry::parseAndCreatePassPipeline` to get the pass manager as an argument rather than creating it inside the function, I think we could easily make it generic to avoid duplicating the pipeline parsing code:
```
class PassRegistry {
  ...
  template <typename PassManagerType>
  void parseAndCreatePassPipeline(PassManagerType &PM, StringRef Pipeline);
};
```

What do you think?

slackito replied:
(and maybe change the `Create` part of the `parseAndCreatePassPipeline` name to something that better describes the new behavior)

vporpo replied:
Hmm then perhaps we should deprecate the PassRegistry in favor of a PassManager that owns the passes?

slackito replied:
We still need something to hold the mapping between names and passes so we can dynamically create pass pipelines based on the flag value. Right now the PassRegistry is that something. What are you proposing exactly?

vporpo replied:
I don't see why we need a parsePassPipeline() in more than one place. Is the PassRegistry's `parsePassPipeline()` still being used after this patch?

slackito replied:
> Is the PassRegistry's `parsePassPipeline()` still being used after this patch?

As it is, no. That's why I proposed templatizing it so we can create pipelines of sandbox IR function passes, region passes, etc with a single parsing function. In the future, if there is more than one sandbox IR function pass, we'll probably want to create arbitrary pipelines out of them too. But I can also just delete that member function if we think we aren't going to need it.

vporpo replied:
Yeah templatizing it is fine, we can do it when needed.

> But I can also just delete that member function if we think we aren't going to need it.

That's what I was thinking of. It seems that we are not going to need it.

Also if the PassRegistry's role ends up being just a mapping from name to pass class, then we could use a .def file.

slackito replied:
How about this?

slackito committed:
Remove PassRegistry, use a .def file for vectorizer RegionPasses.

PassManagers now own the passes they contain, and pipeline parsing is
done by a standalone function in BottomUpVec.cpp.

vporpo commented on llvm/lib/Transforms/Vectorize/SandboxVectorizer/Passes/PassRegistry.def:
```diff
@@ -0,0 +1,20 @@
+#define REGION_PASS(NAME, CREATE_PASS)
+#endif
+
+REGION_PASS("null", NullPass())
```
Also why use `NullPass()` and not just the class name `NullPass` in `REGION_PASS("null", NullPass())`. Isn't the class name easier to work with?

slackito replied:
I just followed the pattern for the existing `FUNCTION_PASS` in llvm/lib/Passes/PassRegistry.def. I have no preference one way or the other.

vporpo commented on llvm/lib/Transforms/Vectorize/SandboxVectorizer/Passes/BottomUpVec.cpp:
```diff
@@ -10,10 +10,69 @@
+}
+
+/// Adds to `RPM` a sequence of region passes described by `Pipeline`.
+static void parseAndCreatePassPipeline(RegionPassManager &RPM,
```
Now that it's the pass managers that own the passes it looks a bit out of place that populating them is done with a static function, because the only users of it are the pass managers.

Wdyt about moving this to the PassManager base class?

slackito replied:
If we do this, then sandboxir::PassManager has to know about vectorizer-specific passes. Is this a desired outcome?

vporpo replied:
I meant making `parseAndCreatePassPipeline` a member function of `PassManager` such that we call:
```
RPM.parseAndCreatePassPipeline(UserDefinedPassPipeline);
```
instead of:
```
parseAndCreatePassPipeline(RPM, UserDefinedPassPipeline);
```

slackito replied:
Right. But `parseAndCreatePassPipeline` needs to know about the existing vectorizer passes in order to instantiate them.

Do you want llvm/lib/SandboxIR/PassManager.cpp to include all the passes from llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Passes/* ? Should we move `RegionPassManager` back into the vectorizer?

vporpo replied:
Well, conceptually parsing the pass pipeline string and building passes should not be a static function in BottomUpVec.cpp because it is a pass-manager task, not a vectorization-specific task.

Oh yes, the creator function would need to have access to the passes. Hmm could we perhaps keep the creator function here in a lambda `createPass(Name)` and pass the lambda as an argument to the parse function?

slackito replied:
Done. Left `createPass` as a static function in BottomUpVec.h instead of a lambda because the includes and defines needed for the .def file messed up indentation and I preferred to keep that ugliness contained in its own function.

I also added a new unit test, similar to the now deleted PassRegistry test, that checks the pipeline parsing function (we also have some very basic coverage from lit tests)

slackito committed:
Added  cleanup in PassRegistry.def

slackito commented on llvm/lib/Transforms/Vectorize/SandboxVectorizer/Passes/PassRegistry.def:
```diff
@@ -0,0 +1,20 @@
+#define REGION_PASS(NAME, CREATE_PASS)
+#endif
+
+REGION_PASS("null", NullPass())
```
Added `#undef REGION_PASS`. By the time we get to the end the macros is always defined, either to something useful by the user or to an empty macro above.

slackito committed:
Moved pipeline creation to a PassManager method.

The new method receives a CreatePass function as an argument to avoid
the PassManager class having to "know" about existing passes in, for
example, SandboxVectorizer.

Also added back a unit test similar to the now deleted PassRegistry test
that checks PassManager::setPassPipeline.

vporpo commented on llvm/unittests/SandboxIR/PassTest.cpp:
```diff
@@ -255,3 +255,58 @@ define i8 @foo(i8 %v0, i8 %v1) {
+  FPM.runOnFunction(*F);
+  EXPECT_EQ(Str, "foobarfoo");
+
+  // Check that a second call to setPassPipeline clears the previous pipeline.
```
Should we allow a second call to setPassPipeline() ? Shouldn't we crash if we try to do so instead?

vporpo reviewed (approved):
LGTM apart from the multiple invocations to setPassPipeline.

slackito committed:
Assert on multiple calls to setPassPipeline. Update unit test.
